### ISSUE TITLE
General Purpose CLI Dashboard

### ISSUE THREAD

**Issue by mikaelbr** [mikaelbr/open-source-ideas#8]
## Project description

This is just a conceptual idea I've got some months ago, and I haven't really thought it through too much, or researched the topic, but I thought I'd share it here to hear other peoples opinion about it.

Weirdly enough, but 2016 seems to be the year the CLI makes it revised return for the fullest. We've seen many tools that have gotten much more focus on usability/UX in CLI, instead of just "good old API-design" for commands. You have interactive watchers such as in Jest, [Webpack Dashboard](https://formidable.com/blog/2016/08/15/introducing-webpack-dashboard/), and more. Things like [blessed](https://github.com/chjj/blessed) (and maybe for some [react-blessed](https://github.com/Yomguithereal/react-blessed)) seems to have allowed people to create more advanced CLI-systems in a easier way.

Why not use this to create a pluggable general purpose CLI dashboard? By this I mean for instance just the ability to refer to different Unix-streams (or commands) and a grid size and automatically create an interactive dashboard from it.

Something like either a hookup file:

```js
// dashboard.js
const dashboard = require('cli-dashboard');
dashboard([
  { command: 'mocha -R dots -w', size: 2 },
  { command: 'webpack -d -w', size: 4 },
  { command: 'onchange 'app/**/*js' -- eslint', size: 2}
]);
```

or by using `package.json` scripts:
```js
// dashboard.js
const dashboard = require('cli-dashboard');
dashboard([
  { script: 'test', size: 2 },
  { script: 'build:js', size: 4 },
  { script: 'watch:lint', size: 2}
]);
```
and running

```shell
node dashboard.js
```

or even

```shell
cli-dashboard
```

Could launch a CLI dashboard looking something like

```
--------------------------------------------------------------------------------
|                                       |                                      |
|                                       |                                      |
|                 mocha                 |                                      |
|                                       |                                      |
|                                       |                                      |
-----------------------------------------                                      |
|                                       |                 webpack              |
|                                       |                                      |
|                eslint                 |                                      |
|                                       |                                      |
|                                       |                                      |
|                                       |                                      |
--------------------------------------------------------------------------------
```

Based on the releative sizes.

Maybe this isn't as useful as I'm thinking now, but maybe the idea sparks some other ideas with you.

## Relevant Technology

JavaScript, Node.js, CLI, blessed, NPM, Frontend build systems 

## Who is this for

Someone with experience with CLI systems and the need to organize different CLI outputs. Labelling this as a lot of work, even though I don't think it has to be that massive.

**Comment by BobWhitelock** [comment]
Maybe I'm missing something, but this sounds like exactly the kind of thing Tmux can be used for. With something like [Tmuxinator](https://github.com/tmuxinator/tmuxinator) the setup you describe could be achieved with just a few lines of declarative config.

**Comment by mikaelbr** [comment]
Interesting. That might very much be 👍 I haven't much experience with tmux. But tmux is for Termals only, though? So it might be for latest Windows with bash, but for many Windows developers this won't work.

**Comment by BobWhitelock** [comment]
That's true, I think Tmux is only for Unix-like OSes so wouldn't work for this use case on older versions of Windows, and I don't know of an alternative that would work on Windows. So it could be interesting for someone to attempt a cross-platform, simplified alternative like this - although I guess the need for it would diminish as people upgrade their Windows versions :).

**Comment by mikaelbr** [comment]
True. Maybe not all Windows developers will transition to *sh terminals, though. Some might even like cmd/PowerShell?

**Comment by BobWhitelock** [comment]
That's true, in which case I hope one of them decides to make this! :)

**Comment by mikaelbr** [comment]
Hehe. I'll definitely check out tmux more. Thanks for the great tip, didn't think about that at all. I'll leave this open some more to see if someone want's to do it for learning more about CLI-development and we'll see if there are any takers.

**Comment by Haroenv** [comment]
https://github.com/yaronn/blessed-contrib also exists

**Comment by mikaelbr** [comment]
Cool. That's looks pretty useful, and now that I see it I know I've seen it before at some point. It's not exactly the same as I envisioned, as it's more low level, but looks very interesting non the less.
